### EDA E-commerce Customer Behavior Analysis

### Problem Statement

This project aims to analyze customer behavior in an e-commerce platform by examining factors such as demographic attributes (age, gender, location), engagement metrics (login frequency, session duration, pages per session), and purchasing behavior (cart abandonment, wishlist activity, lifetime value), in order to identify patterns related to customer retention, high-value users, and churn.

### Data Collection

* Dataset Source: https://www.kaggle.com/datasets/dhairyajeetsingh/ecommerce-customer-behavior-dataset

* The data consists of 25 columns and 50,000 rows.

3. DataSet Information

* Age : Age of the customer
* Gender : Gender of the customer (Male/Female)
* Country : Country where the customer is located
* City : City of the customer
* Membership_Years : Number of years the customer has been using the platform
* Login_Frequency : Number of times the customer logs into the platform
* Session_Duration_Avg : Average time spent per session
* Pages_Per_Session : Average number of pages viewed per session
* Cart_Abandonment_Rate : Percentage of times a customer adds items to cart but does not complete purchase
* Wishlist_Items : Number of items added to wishlist
* Email_Open_Rate : Percentage of marketing emails opened by the customer
* Customer_Service_Calls : Number of times the customer contacted support
* Product_Reviews_Written : Number of reviews written by the customer
* Social_Media_Engagement_Score : Level of interaction on social media platforms
* Mobile_App_Usage : Usage level of the mobile application
* Payment_Method_Diversity : Number of different payment methods used
* Lifetime_Value : Total revenue generated from the customer
* Credit_Balance : Available credit balance of the customer
* Churned : Whether the customer has stopped using the platform (0 = No, 1 = Yes)
* Signup_Quarter : Quarter in which the customer signed up (Q1, Q2, Q3, Q4)

Step 1: Import all the required libraries

In [2]:
# Importing required libraries

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')


Step 2: Read the dataset

In [3]:
df = pd.read_csv('ecommerce_customer_dataset.csv')
df.head()

,Age,Gender,Country,City,Membership_Years,Login_Frequency,Session_Duration_Avg,Pages_Per_Session,Cart_Abandonment_Rate,Wishlist_Items,...,Email_Open_Rate,Customer_Service_Calls,Product_Reviews_Written,Social_Media_Engagement_Score,Mobile_App_Usage,Payment_Method_Diversity,Lifetime_Value,Credit_Balance,Churned,Signup_Quarter
0,43.0,Male,France,Marseille,2.9,14.0,27.4,6.0,50.6,3.0,...,17.9,9.0,4.0,16.3,20.8,1.0,953.33,2278.0,0,Q1
1,36.0,Male,UK,Manchester,1.6,15.0,42.7,10.3,37.7,1.0,...,42.8,7.0,3.0,NaN,23.3,3.0,1067.47,3028.0,0,Q4
2,45.0,Female,Canada,Vancouver,2.9,10.0,24.8,1.6,70.9,1.0,...,0.0,4.0,1.0,NaN,8.8,NaN,1289.75,2317.0,0,Q4
3,56.0,Female,USA,New York,2.6,10.0,38.4,14.8,41.7,9.0,...,41.4,2.0,5.0,85.9,31.0,3.0,2340.92,2674.0,0,Q1
4,35.0,Male,India,Delhi,3.1,29.0,51.4,NaN,19.1,9.0,...,37.9,1.0,11.0,83.0,50.4,4.0,3041.29,5354.0,0,Q4


In [4]:
df.shape

(50000, 25)

## Data Checks to Perform

* Check for missing values in the dataset
* Check for duplicate records and remove if necessary
* Verify data types of each column
* Check the number of unique values in each column
* Perform statistical summary of numerical features
* Analyze different categories in categorical columns
* Check the distribution of the target variable (Churned)
* Identify outliers in numerical features
* Validate data ranges and consistency of values
* Analyze distribution of key features using plots
* Examine relationships between features using correlation analysis

### 1. Handle Missing Values

*  checking the missing value in dataset


In [ ]:

missing_values = df.isnull().sum()
missing_percentage = (missing_values / len(df)) * 100

missing_df = pd.DataFrame({'missing_value' : missing_values,
                           'percentage'  : missing_percentage})

print(missing_df)

                               missing_value  percentage
Age                                     2495       4.990
Gender                                     0       0.000
Country                                    0       0.000
City                                       0       0.000
Membership_Years                           0       0.000
Login_Frequency                            0       0.000
Session_Duration_Avg                    3399       6.798
Pages_Per_Session                       3000       6.000
Cart_Abandonment_Rate                      0       0.000
Wishlist_Items                          4000       8.000
Total_Purchases                            0       0.000
Average_Order_Value                        0       0.000
Days_Since_Last_Purchase                3000       6.000
Discount_Usage_Rate                     3500       7.000
Returns_Rate                            4491       8.982
Email_Open_Rate                         2528       5.056
Customer_Service_Calls         

* Handle missing values

1. Columns to Fill with Median
* Continuous numeric columns where missing = unknown value:
    * Age → median
    * Session_Duration_Avg → median
    * Pages_Per_Session → median
    * Days_Since_Last_Purchase → median

In [11]:
num_median_cols = ['Age', 'Session_Duration_Avg', 'Pages_Per_Session', 'Days_Since_Last_Purchase']

for col in num_median_cols:
    df[col] = df[col].fillna(df[col].median())

2. Columns to fill with 0

* Missing likely means “didn’t do it / no activity”:

    * Wishlist_Items
    * Discount_Usage_Rate
    * Returns_Rate
    * Email_Open_Rate
    * Customer_Service_Calls
    * Product_Reviews_Written
    * Mobile_App_Usage
    * Social_Media_Engagement_Score
    * Credit_Balance



In [12]:
zero_fill_cols = ['Wishlist_Items','Discount_Usage_Rate','Returns_Rate','Email_Open_Rate','Customer_Service_Calls','Product_Reviews_Written','Mobile_App_Usage','Social_Media_Engagement_Score','Credit_Balance']

for col in zero_fill_cols:
    df[col] = df[col].fillna(0)

3. Categorical / Small Missing Columns

* Payment_Method_Diversity → fill with median or mode (usually small % missing, 5%)

In [13]:
df['Payment_Method_Diversity'] = df['Payment_Method_Diversity'].fillna(df['Payment_Method_Diversity'].median())

In [14]:
# checking missing values again

df.isna().sum()

Age                              0
Gender                           0
Country                          0
City                             0
Membership_Years                 0
Login_Frequency                  0
Session_Duration_Avg             0
Pages_Per_Session                0
Cart_Abandonment_Rate            0
Wishlist_Items                   0
Total_Purchases                  0
Average_Order_Value              0
Days_Since_Last_Purchase         0
Discount_Usage_Rate              0
Returns_Rate                     0
Email_Open_Rate                  0
Customer_Service_Calls           0
Product_Reviews_Written          0
Social_Media_Engagement_Score    0
Mobile_App_Usage                 0
Payment_Method_Diversity         0
Lifetime_Value                   0
Credit_Balance                   0
Churned                          0
Signup_Quarter                   0
dtype: int64

In [15]:
df.notna().sum()

Age                              50000
Gender                           50000
Country                          50000
City                             50000
Membership_Years                 50000
Login_Frequency                  50000
Session_Duration_Avg             50000
Pages_Per_Session                50000
Cart_Abandonment_Rate            50000
Wishlist_Items                   50000
Total_Purchases                  50000
Average_Order_Value              50000
Days_Since_Last_Purchase         50000
Discount_Usage_Rate              50000
Returns_Rate                     50000
Email_Open_Rate                  50000
Customer_Service_Calls           50000
Product_Reviews_Written          50000
Social_Media_Engagement_Score    50000
Mobile_App_Usage                 50000
Payment_Method_Diversity         50000
Lifetime_Value                   50000
Credit_Balance                   50000
Churned                          50000
Signup_Quarter                   50000
dtype: int64

### Insight or Observation

* Out of 25 columns, 14 columns have missing values.
* Most missing values are less than 10% of the total data(50,000 rows), so we can safely fill them without dropping columns.
* Columns like Social_Media_Engagement_Score (12%) and Credit_balance (11%) have slightly higher missing values, but still managable.
* Missing values can be handled by:
    * FIlling continuous numeric columns with median
    * Filling activity/count columns with 0
    * Filling categorical columns with mode

### 2. Check for duplicate records and remove if necessary

In [16]:
df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
49995    False
49996    False
49997    False
49998    False
49999    False
Length: 50000, dtype: bool

In [17]:
df.duplicated().sum()

np.int64(0)

### Insights/Observation

* There are no dublicate values in the dataset.

### 3. Verify data types of each column


In [18]:
df.dtypes

Age                              float64
Gender                               str
Country                              str
City                                 str
Membership_Years                 float64
Login_Frequency                  float64
Session_Duration_Avg             float64
Pages_Per_Session                float64
Cart_Abandonment_Rate            float64
Wishlist_Items                   float64
Total_Purchases                  float64
Average_Order_Value              float64
Days_Since_Last_Purchase         float64
Discount_Usage_Rate              float64
Returns_Rate                     float64
Email_Open_Rate                  float64
Customer_Service_Calls           float64
Product_Reviews_Written          float64
Social_Media_Engagement_Score    float64
Mobile_App_Usage                 float64
Payment_Method_Diversity         float64
Lifetime_Value                   float64
Credit_Balance                   float64
Churned                            int64
Signup_Quarter  

### Data Type Observation
* The dataset has 25 columns.
* Numeric columns (float64 / int64):
* Continuous / measurable features: Age, Membership_Years, Login_Frequency, Session_Duration_Avg, Pages_Per_Session, Cart_Abandonment_Rate, Wishlist_Items, Total_Purchases, Average_Order_Value, Days_Since_Last_Purchase, Discount_Usage_Rate, Returns_Rate, Email_Open_Rate, Customer_Service_Calls, Product_Reviews_Written, Social_Media_Engagement_Score, Mobile_App_Usage, Payment_Method_Diversity, Lifetime_Value, Credit_Balance
Target variable: Churned (int64)
* Categorical columns (object / str):
Gender, Country, City, Signup_Quarter
* This separation helps decide:
* Numeric columns → analyze distributions, correlations, outliers
* Categorical columns → analyze counts, frequencies, and segmentations

### 3: Check for duplicates, unique values, and basic statistics.

* Purpose: Remove rows that are exactly the same → avoid double counting in analysis

In [ ]:
# 1. check duplicate rows

duplicates_count = df.duplicated().sum()

print(f"Number of duplicate rows: {duplicates_count}")


Number of duplicate rows: 0


In [20]:
# Remove duplicates if any
df = df.drop_duplicates()

### Observation

* Number of duplicate rows found: 0
* Dropped duplicates to ensure clean data.


### 2. Check Unique Values in Each Column
* Purpose: Understand how many distinct categories exist in categorical columns

In [24]:
for col in df.select_dtypes(include='str').columns:
    print(f"{col} : {df[col].nunique()} unique values")

Gender : 3 unique values
Country : 8 unique values
City : 40 unique values
Signup_Quarter : 4 unique values


In [22]:
df.nunique()

Age                                 62
Gender                               3
Country                              8
City                                40
Membership_Years                   200
Login_Frequency                     46
Session_Duration_Avg               658
Pages_Per_Session                  224
Cart_Abandonment_Rate             1000
Wishlist_Items                      25
Total_Purchases                    150
Average_Order_Value              16407
Days_Since_Last_Purchase           235
Discount_Usage_Rate               1728
Returns_Rate                       394
Email_Open_Rate                    720
Customer_Service_Calls              21
Product_Reviews_Written             18
Social_Media_Engagement_Score      972
Mobile_App_Usage                   540
Payment_Method_Diversity             5
Lifetime_Value                   45447
Credit_Balance                    5234
Churned                              2
Signup_Quarter                       4
dtype: int64

In [23]:
df['Gender'].unique()

<StringArray>
['Male', 'Female', 'Other']
Length: 3, dtype: str

In [25]:
df['Country'].unique()

<StringArray>
['France', 'UK', 'Canada', 'USA', 'India', 'Japan', 'Germany', 'Australia']
Length: 8, dtype: str

In [26]:
df['City'].unique()

<StringArray>
[  'Marseille',  'Manchester',   'Vancouver',    'New York',       'Delhi',
       'Tokyo',      'Berlin',     'Houston',     'Calgary',     'Glasgow',
      'Munich',     'Phoenix',       'Paris',      'Ottawa', 'Los Angeles',
   'Bangalore',    'Adelaide',    'Yokohama',      'Sydney',      'Nagoya',
   'Frankfurt',     'Cologne',     'Hamburg',        'Nice',     'Chicago',
    'Toulouse',    'Brisbane',       'Leeds',      'Mumbai',    'Montreal',
   'Melbourne',       'Osaka',        'Lyon',   'Hyderabad',     'Chennai',
  'Birmingham',      'London',     'Toronto',       'Perth',       'Kyoto']
Length: 40, dtype: str

In [27]:
df['Signup_Quarter'].unique()

<StringArray>
['Q1', 'Q4', 'Q2', 'Q3']
Length: 4, dtype: str

### Observation

* Gender -> 3(Male, Female, Others)
* country -> 8 ['France', 'UK', 'Canada', 'USA', 'India', 'Japan', 'Germany', 'Australia']
* City -> 40 [  'Marseille',  'Manchester',   'Vancouver',    'New York',       'Delhi',
       'Tokyo',      'Berlin',     'Houston',     'Calgary',     'Glasgow',
      'Munich',     'Phoenix',       'Paris',      'Ottawa', 'Los Angeles',
   'Bangalore',    'Adelaide',    'Yokohama',      'Sydney',      'Nagoya',
   'Frankfurt',     'Cologne',     'Hamburg',        'Nice',     'Chicago',
    'Toulouse',    'Brisbane',       'Leeds',      'Mumbai',    'Montreal',
   'Melbourne',       'Osaka',        'Lyon',   'Hyderabad',     'Chennai',
  'Birmingham',      'London',     'Toronto',       'Perth',       'Kyoto']

* Signup_Quarter -> 4 ['Q1', 'Q4', 'Q2', 'Q3']


### 3. Basic Statistics for Numeric Columns
* Purpose: See distributions, min/max, mean, std → identify outliers or weird values

In [28]:
df.describe()

,Age,Membership_Years,Login_Frequency,Session_Duration_Avg,Pages_Per_Session,Cart_Abandonment_Rate,Wishlist_Items,Total_Purchases,Average_Order_Value,Days_Since_Last_Purchase,...,Returns_Rate,Email_Open_Rate,Customer_Service_Calls,Product_Reviews_Written,Social_Media_Engagement_Score,Mobile_App_Usage,Payment_Method_Diversity,Lifetime_Value,Credit_Balance,Churned
count,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,...,50000.000000,50000.000000,50000.00000,50000.000000,50000.000000,50000.000000,50000.00000,50000.000000,50000.000000,50000.000000
mean,37.812800,2.984009,11.624660,27.602240,8.717542,57.079973,3.954520,13.111576,123.117330,29.265300,...,6.080833,19.879356,5.66274,2.653580,25.840730,17.434446,2.33618,1440.626292,1749.947600,0.289000
std,11.535688,2.059105,7.810657,10.497234,3.663996,16.282723,3.274206,7.017312,175.569714,28.866043,...,6.164110,14.625661,2.69171,2.361001,21.530263,10.659451,1.08465,907.249443,1309.277071,0.453302
min,5.000000,0.100000,0.000000,1.000000,1.000000,0.000000,0.000000,-13.000000,26.380000,0.000000,...,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,1.00000,0.000000,0.000000,0.000000
25%,30.000000,1.400000,6.000000,20.200000,6.200000,46.400000,1.000000,8.000000,87.050000,9.000000,...,2.100000,8.000000,4.00000,1.000000,5.700000,10.200000,2.00000,789.817500,647.000000,0.000000
50%,38.000000,2.500000,11.000000,26.800000,8.400000,58.100000,3.000000,12.000000,112.970000,21.000000,...,4.900000,18.800000,5.00000,2.000000,23.700000,17.200000,2.00000,1243.415000,1697.000000,0.000000
75%,45.000000,4.000000,17.000000,34.000000,11.000000,68.700000,6.000000,17.000000,144.440000,39.000000,...,8.600000,29.700000,7.00000,4.000000,40.700000,24.600000,3.00000,1874.000000,2664.000000,1.000000
max,200.000000,10.000000,46.000000,75.600000,24.100000,143.743350,28.000000,128.700000,9666.379178,287.000000,...,99.615734,91.700000,21.00000,21.000000,100.000000,61.900000,5.00000,8987.240000,7197.000000,1.000000


### Observation from basic statistics

* 1. Age
Mean ~38, median 38 → distribution is fairly centered
Min = 5, Max = 200 → Max is suspicious (likely outlier or data entry error)
* 2. Membership_Years
Range 0.1 → 10 years, mean ~3 → most customers are relatively new
25–75% percentile: 1.4 → 4 years → skewed towards newer members
* 3. Login_Frequency & Session_Duration_Avg
Login_Frequency: 0 → 46 logins, mean ~12 → some extreme high users
Session_Duration_Avg: 1 → 75 minutes, mean ~27 → some outliers (very long sessions)
* 4. Pages_Per_Session
1 → 24 pages, median 8.4 → some heavy users
* 5. Cart_Abandonment_Rate
Mean ~57%, 75% percentile ~68% → high abandonment common
* 6.Wishlist_Items, Total_Purchases, Average_Order_Value
Wishlist_Items max 28 → some very active users
Total_Purchases negative value (-13) → data error (needs cleaning)
Average_Order_Value max 9666 → huge outlier (could skew revenue analysis)
* 7. Days_Since_Last_Purchase
Min 0 → max 287 days, mean ~29 → some inactive customers
* 8. Behavioral / Activity Columns
Discount_Usage_Rate, Returns_Rate, Email_Open_Rate, Customer_Service_Calls, Product_Reviews_Written, Social_Media_Engagement_Score, Mobile_App_Usage, Payment_Method_Diversity → all have some extreme high users, check for outliers
* 9.Financial Metrics
Lifetime_Value: mean ~1440, max 8987 → a few high-value customers dominate
Credit_Balance: mean ~1750, max 7197 → some very high balances
* 10.Target Variable (Churned)
0 → 71%, 1 → 29% → slightly imbalanced, may need stratification if building models